pkg/flavor/flavor.go

```diff
@@ -44,6 +53,263 @@ func NewMethods(clientset kubecli.KubevirtClient) Methods {
+	}
+
+	revisionPatch, err := json.Marshal(specRevision)
+	if err != nil {
+		return nil, err
+	}
+
+	return &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            revisionName,
+			Namespace:       vm.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(vm, virtv1.VirtualMachineGroupVersionKind)},
+		},
+		Data: runtime.RawExtension{Raw: revisionPatch},
+	}, nil
+
+}
+
+func (m *methods) createPreferenceRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+
+	switch strings.ToLower(vm.Spec.Preference.Kind) {
+	case apiflavor.SingularPreferenceResourceName, apiflavor.PluralPreferenceResourceName:
+		preference, err := m.findPreference(vm)
+		if err != nil {
+			return nil, err
+		}
+
+		revision, err := CreatePreferenceControllerRevision(vm, GetRevisionName(vm.Name, preference.Name, preference.UID, preference.Generation), preference.APIVersion, &preference.Spec)
+		if err != nil {
+			return nil, err
+		}
+		return revision, nil
+	case apiflavor.ClusterSingularPreferenceResourceName, apiflavor.ClusterPluralPreferenceResourceName:
+		clusterPreference, err := m.findClusterPreference(vm)
+		if err != nil {
+			return nil, err
+		}
+
+		revision, err := CreatePreferenceControllerRevision(vm, GetRevisionName(vm.Name, clusterPreference.Name, clusterPreference.UID, clusterPreference.Generation), clusterPreference.APIVersion, &clusterPreference.Spec)
+		if err != nil {
+			return nil, err
+		}
+		return revision, nil
+	default:
+		return nil, fmt.Errorf("got unexpected kind in PreferenceMatcher: %s", vm.Spec.Preference.Kind)
+	}
+}
+
+func (m *methods) storePreferenceRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+
+	if vm.Spec.Preference == nil || len(vm.Spec.Preference.RevisionName) > 0 {
+		return nil, nil
+	}
+
+	preferenceRevision, err := m.createPreferenceRevision(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = m.clientset.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), preferenceRevision, metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			// Grab the existing revision to check the data it contains
+			existingRevision, err := m.clientset.AppsV1().ControllerRevisions(vm.Namespace).Get(context.Background(), preferenceRevision.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			// If the data between the two differs return an error, otherwise continue and store the name below.
+			if bytes.Compare(existingRevision.Data.Raw, preferenceRevision.Data.Raw) != 0 {
+				return nil, fmt.Errorf("found existing ControllerRevision with unexpected data: %s", preferenceRevision.Name)
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	vm.Spec.Preference.RevisionName = preferenceRevision.Name
+
+	return preferenceRevision, nil
+
+}
+
+func GenerateRevisionNamePatch(flavorRevision, preferenceRevision *appsv1.ControllerRevision) ([]byte, error) {
```

Comment:
does this need to be an exported function name?

Reply:
This has just to help keep the tests clean, happy to address in a follow up if it's an issue.

https://github.com/kubevirt/kubevirt/blob/f340573749fa281b5259fca409ef2fe99b297a77/pkg/flavor/flavor_test.go#L160-L161